query/bridges.go

```diff
@@ -23,35 +22,76 @@ func (b QueryServiceBridge) Query(ctx context.Context, req *Request) (flux.Resul
 	return flux.NewResultIteratorFromQuery(query), nil
 }
 
-// ProxyQueryServiceBridge implements ProxyQueryService while consuming a QueryService interface.
-type ProxyQueryServiceBridge struct {
-	QueryService QueryService
+// QueryServiceProxyBridge implements QueryService while consuming a ProxyQueryService interface.
+type QueryServiceProxyBridge struct {
+	ProxyQueryService ProxyQueryService
 }
 
-func (b ProxyQueryServiceBridge) Query(ctx context.Context, w io.Writer, req *ProxyRequest) (int64, error) {
-	results, err := b.QueryService.Query(ctx, &req.Request)
-	if err != nil {
-		return 0, err
+func (b QueryServiceProxyBridge) Query(ctx context.Context, req *Request) (flux.ResultIterator, error) {
+	d := csv.Dialect{ResultEncoderConfig: csv.DefaultEncoderConfig()}
+	preq := &ProxyRequest{
+		Request: *req,
+		Dialect: d,
 	}
-	defer results.Release()
 
-	// Setup headers
-	if w, ok := w.(http.ResponseWriter); ok {
-		w.Header().Set("Trailer", "Influx-Query-Statistics")
-	}
+	r, w := io.Pipe()
+	statsChan := make(chan flux.Statistics, 1)
 
-	encoder := req.Dialect.Encoder()
-	n, err := encoder.Encode(w, results)
+	go func() {
+		var stats flux.Statistics
+		var err error
+		defer func() {
+			_ = w.CloseWithError(err)
+			statsChan <- stats
+		}()
```

Comment:
Why is this a defer instead of following the call to Query?

Reply:
If the call to `Query` panics somewhere, we may not close the pipe and write the stats, is that right?  I wanted to make sure we do that even if a panic occurs.

Reply:
If a panic occurs it will crash the whole queryd process. Since a new go routine was created if that go routine panics then the process will exit. In that case there is no real need to clean up the pipe. I think we can ignore panics here as it means a complete process shutdown.